### PR TITLE
[RISCV] Rematerialize vid.v

### PR DESCRIPTION
**Pull request by lukel97**
This adds initial support for rematerializing vector instructions, starting with vid.v since it's simple and has the least number of operands. It has one passthru operand which we need to check is undefined. It also has an AVL operand, but it's fine to rematerialize with it because it's scalar and register allocation is split between vector and scalar.
    
RISCVInsertVSETVLI can still happen before vector regalloc if -riscv-vsetvl-after-rvv-regalloc is false, so this makes sure that we only rematerialize after regalloc by checking for the implicit uses that are added.

### CODE REVIEW

**Review by topperc — APPROVED** [review]
LGTM

File: llvm/test/CodeGen/RISCV/rvv/remat.ll [hunk 1]
```diff
@@ -0,0 +1,111 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
```
Comment (wangpc-pp) [inline llvm/test/CodeGen/RISCV/rvv/remat.ll:1]: Rematerializable instructions can be hoisted in MachineLICM, need we add some loop tests here?

**Review by wangpc-pp — APPROVED** [review]
LGTM.

File: llvm/test/CodeGen/RISCV/rvv/remat.ll (hunk 1 above)
Comment (topperc) [reply llvm/test/CodeGen/RISCV/rvv/remat.ll:1]: LICM should be able to hoist vid.v even it wasn't rematerializable. Right?

Comment (wangpc-pp) [reply llvm/test/CodeGen/RISCV/rvv/remat.ll:1]: Yeah, you're right!

**Pull request merged**